### PR TITLE
[Fix] NetworkContext defined differently in compilation units

### PR DESCRIPTION
**Pull request by danewalton**
- The `NetworkContext` for ESP was being defined differently than the main sample file which could cause some memory corruption when values were dereferenced. 

https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/041be65e226c98d24f74747c2e6aab53c7ae8f8c/demos/sample_azure_iot_adu/sample_azure_iot_adu.c#L159-L163

**VS**

https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/041be65e226c98d24f74747c2e6aab53c7ae8f8c/demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_tls_esp32.c#L41-L50 

- This update follows the FreeRTOS recommended practice of putting pParams as a `void` pointer if you plan on having different transports being passed: [LINK](https://github.com/FreeRTOS/FreeRTOS/blob/dc2c031c451744853dfc12aa94e0a5438c56bde5/FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/Common/http_demo_utils.c#L69).

Relevant GH Issues: https://github.com/Azure-Samples/iot-middleware-freertos-samples/issues/256 | https://github.com/Azure-Samples/iot-middleware-freertos-samples/issues/254

### CODE REVIEW

**Comment by danewalton** [comment]
This is working for L475 and ESP32 for me.

File: demos/common/transport/transport_socket.c [hunk 1]
```diff
@@ -33,15 +33,13 @@ extern void vLoggingPrintf( const char * pcFormatString,
...
-    SSLContextHandle xSSLContext;
-} TlsTransportParams_t;
-
+/* Each transport defines the same NetworkContext. The user then passes their respective transport */
+/* as pParams for the transport which is defined in the transport header file */
+/* (here it's SocketTransportParams_t) */
 struct NetworkContext
 {
-    TlsTransportParams_t * pParams;
+    /* SocketTransportParams_t */
```
Comment (ericwol-msft) [inline demos/common/transport/transport_socket.c:41]:
```suggestion
    /* Parameters for the socker transport (SocketTransportParams_t) */
```

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c [hunk 2]
```diff
@@ -71,11 +77,27 @@ SocketTransportStatus_t Azure_Socket_Connect( NetworkContext_t * pNetworkContext
...
 
-    pNetworkContext->xTransport = esp_transport_tcp_init( );
-    pNetworkContext->ulReceiveTimeoutMs = ulReceiveTimeoutMs;
-    pNetworkContext->ulSendTimeoutMs = ulSendTimeoutMs;
+    SocketTransportParams_t * pxSocketTransport = (SocketTransportParams_t *)pNetworkContext->pParams;
 
-    if ( esp_transport_connect( pNetworkContext->xTransport, pHostName, usPort, ulReceiveTimeoutMs ) < 0 )
+    if((pxSocketTransport == NULL))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pNetworkContext->pParams=%p.", pxSocketTransport );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c:84]: we already know this is null, why print it?

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c [hunk 3]
```diff
@@ -113,11 +136,22 @@ void Azure_Socket_Close( NetworkContext_t * pNetworkContext )
         return;
     }
 
+    SocketTransportParams_t * pxSocketTransport = (SocketTransportParams_t *)pNetworkContext->pParams;
+
+    if((pxSocketTransport == NULL))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pNetworkContext->pParams=%p.", pxSocketTransport );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c:143]: same, no need to print this

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c [hunk 4]
```diff
@@ -136,7 +170,23 @@ int32_t Azure_Socket_Recv( NetworkContext_t * pNetworkContext,
         return eSocketTransportInvalidParameter;
     }
 
-    lsocketStatus = esp_transport_read( pNetworkContext->xTransport, pBuffer, xBytesToRecv, pNetworkContext->ulReceiveTimeoutMs );
+    SocketTransportParams_t * pxSocketTransport = (SocketTransportParams_t *)pNetworkContext->pParams;
+
+    if((pxSocketTransport == NULL))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pNetworkContext->pParams=%p.", pxSocketTransport );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c:177]: no need to print NULL values

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c [hunk 5]
```diff
@@ -136,7 +170,23 @@ int32_t Azure_Socket_Recv( NetworkContext_t * pNetworkContext,
...
+        return eSocketTransportInvalidParameter;
+    }
+
+    EspSocketTransportParams_t * pxEspSocketTransport = (EspSocketTransportParams_t*)pxSocketTransport->xSocketContext;
+
+    if((pxEspSocketTransport == NULL))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxSocketTransport->xSocketContext=%p.", pxEspSocketTransport );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c:185]: no need to print NULL values

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c [hunk 6]
```diff
@@ -162,7 +212,23 @@ int32_t Azure_Socket_Send( NetworkContext_t * pNetworkContext,
         return eSocketTransportInvalidParameter;
     }
 
-    lsocketStatus = esp_transport_write( pNetworkContext->xTransport, pBuffer, xBytesToSend, pNetworkContext->ulSendTimeoutMs );
+    SocketTransportParams_t * pxSocketTransport = (SocketTransportParams_t *)pNetworkContext->pParams;
+
+    if((pxSocketTransport == NULL))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pNetworkContext->pParams=%p.", pxSocketTransport );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c:219]: no need to print NULL values

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c [hunk 7]
```diff
@@ -162,7 +212,23 @@ int32_t Azure_Socket_Send( NetworkContext_t * pNetworkContext,
...
+        return eSocketTransportInvalidParameter;
+    }
+
+    EspSocketTransportParams_t * pxEspSocketTransport = (EspSocketTransportParams_t*)pxSocketTransport->xSocketContext;
+
+    if((pxEspSocketTransport == NULL))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxSocketTransport->xSocketContext=%p.", pxEspSocketTransport );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_socket_esp32.c:227]: no need to print NULL values

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_tls_esp32.c [hunk 8]
```diff
@@ -139,11 +166,29 @@ void TLS_Socket_Disconnect( NetworkContext_t * pNetworkContext )
         return;
     }
 
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_tls_esp32.c:173]: no need to print NULL values

File: demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_tls_esp32.c [hunk 9]
```diff
@@ -139,11 +166,29 @@ void TLS_Socket_Disconnect( NetworkContext_t * pNetworkContext )
...
+        return;
+    }
+
+    EspTlsTransportParams_t * pxEspTlsTransport = (EspTlsTransportParams_t *)pxTlsParams->xSSLContext;
+
+    if((pxEspTlsTransport == NULL))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams->xSSLContext=%p.", pxEspTlsTransport );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/adu/components/sample-azure-iot/transport_tls_esp32.c:181]: no need to print NULL values

File: demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c [hunk 10]
```diff
@@ -73,35 +80,53 @@ TlsTransportStatus_t TLS_Socket_Connect( NetworkContext_t * pNetworkContext,
...
     }
 
-    pNetworkContext->xTransport = esp_transport_ssl_init( );
-    pNetworkContext->ulReceiveTimeoutMs = ulReceiveTimeoutMs;
-    pNetworkContext->ulSendTimeoutMs = ulSendTimeoutMs;
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c:87]: no need to print NULL values

File: demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c [hunk 11]
```diff
@@ -73,35 +80,53 @@ TlsTransportStatus_t TLS_Socket_Connect( NetworkContext_t * pNetworkContext,
...
+        return eTLSTransportInvalidParameter;
+    }
+
+    EspTlsTransportParams_t * pxEspTlsTransport = (EspTlsTransportParams_t*) pvPortMalloc(sizeof(EspTlsTransportParams_t));
+
+    if(pxEspTlsTransport == NULL)
+    {
+      return eTLSTransportInvalidParameter;
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c:95]: out of memory

File: demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c [hunk 12]
```diff
@@ -139,11 +165,22 @@ void TLS_Socket_Disconnect( NetworkContext_t * pNetworkContext )
         return;
     }
 
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c:172]: no need to print NULL values

File: demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c [hunk 13]
```diff
@@ -162,7 +199,17 @@ int32_t TLS_Socket_Recv( NetworkContext_t * pNetworkContext,
         return eTLSTransportInvalidParameter;
     }
 
-    tlsStatus = esp_transport_read( pNetworkContext->xTransport, pBuffer, xBytesToRecv, pNetworkContext->ulReceiveTimeoutMs );
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c:206]: no need to print NULL values

File: demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c [hunk 14]
```diff
@@ -188,7 +235,17 @@ int32_t TLS_Socket_Send( NetworkContext_t * pNetworkContext,
         return eTLSTransportInvalidParameter;
     }
 
-    tlsStatus = esp_transport_write( pNetworkContext->xTransport, pBuffer, xBytesToSend, pNetworkContext->ulSendTimeoutMs );
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c:242]: no need to print NULL values

File: demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c [hunk 15]
```diff
@@ -251,34 +259,52 @@ TlsTransportStatus_t TLS_Socket_Connect( NetworkContext_t * pNetworkContext,
...
     }
 
-    pNetworkContext->xTransport = esp_transport_ssl_init( );
-    pNetworkContext->ulReceiveTimeoutMs = ulReceiveTimeoutMs;
-    pNetworkContext->ulSendTimeoutMs = ulSendTimeoutMs;
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c:266]: no need to print NULL values

File: demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c [hunk 16]
```diff
@@ -341,11 +368,22 @@ void TLS_Socket_Disconnect( NetworkContext_t * pNetworkContext )
         return;
     }
 
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c:375]: no need to print NULL values

File: demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c [hunk 17]
```diff
@@ -364,7 +402,17 @@ int32_t TLS_Socket_Recv( NetworkContext_t * pNetworkContext,
         return eTLSTransportInvalidParameter;
     }
 
-    tlsStatus = esp_transport_read( pNetworkContext->xTransport, pBuffer, xBytesToRecv, pNetworkContext->ulReceiveTimeoutMs );
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c:409]: no need to print NULL values

File: demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c [hunk 18]
```diff
@@ -390,7 +438,17 @@ int32_t TLS_Socket_Send( NetworkContext_t * pNetworkContext,
         return eTLSTransportInvalidParameter;
     }
 
-    tlsStatus = esp_transport_write( pNetworkContext->xTransport, pBuffer, xBytesToSend, pNetworkContext->ulSendTimeoutMs );
+    TlsTransportParams_t * pxTlsParams = (TlsTransportParams_t*)pNetworkContext->pParams;
+
+    if (( pxTlsParams == NULL ))
+    {
+        ESP_LOGE( TAG, "Invalid input parameter(s): Arguments cannot be NULL. pxTlsParams=%p.", pxTlsParams );
```
Comment (ericwol-msft) [inline demos/projects/ESPRESSIF/esp32/components/sample-azure-iot/transport_tls_esp32.c:445]: no need to print NULL values

File: demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c (hunk 11 above)
Comment (ericwol-msft) [reply demos/projects/ESPRESSIF/aziotkit/components/sample-azure-iot/transport_tls_esp32.c:95]: eTLSTransportInsufficientMemory

**Review by ericwol-msft — APPROVED** [review]

**Comment by danewalton** [comment]
Working using ESP IDF V4.3.4 on Windows

**Pull request merged**